fix(eslint-plugin): [no-unsafe-type-assertion] fix for unsafe assertion to a constrained type parameter

controversial opened the pull request:
## PR Checklist

- [x] Addresses an existing open issue: fixes #10453 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Attempts to fix the issue described in #10453:
```ts
type Obj = { foo: string };
function func<T extends Obj>() {
  const o: Obj = { foo: 'hi' };
  // this type assertion is unsafe because T could be instantiated with a more specific subtype of Obj
  // no-unsafe-type-assertion should (but didn’t) report this
  const myObj = o as T;
}
```

After experimenting with the logic in `no-unsafe-type-assertion`, I found that obtaining the “types to check” using `services.getTypeAtLocation` in place of `getConstrainedTypeAtLocation` (i.e. omitting a call to `getBaseConstraintOfType`) allows the subsequent `isTypeAssignableTo` check to correctly catch this case.

Additionally, I added a new error message for this case: if the “expression type” is *not* assignable to the “asserted type”, but the asserted type has a constraint to which the expression type is assignable, the rule will report the following:
> Unsafe type assertion: the original type is assignable to the constraint of type '{{type}}', but '{{type}}' could be instantiated with a different subtype of its constraint.

---

This is my first contribution to this project and I’m broadly unfamiliar with both ESLint plugin development and TypeScript internals, so it’s very possible that I’m missing a hidden implication of this change—i.e. I’m not sure if there was an important reason to use `getConstrainedTypeAtLocation` in the original implementation. However, the existing tests for this rule are still passing.

controversial commented:
With respect to a custom error message, @kirkwaiblinger suggested:
> As a first pass, I think we could just model it after the TS error message?
> 
> ```
> Unsafe type assertion: 'Object' is assignable to the constraint of type 'CustomObjectT', but 'CustomObjectT' could be instantiated with a different subtype of constraint 'Object'.
> ```

In the original PR of this rule, #10051, a discussion about the appropriate verbosity of the error message [resolved](https://github.com/typescript-eslint/typescript-eslint/pull/10051#discussion_r1795639908) to omit any types other than the *explicitly written type* from the error message in order to prevent excessively verbose errors.

With that in mind, I’m starting out with the following error message:

> Unsafe type assertion: **the original type** is assignable to the constraint of type 'CustomObjectT', but 'CustomObjectT' could be instantiated with a different subtype of **the constraint**.

I’m happy to amend this message to be more verbose if we decide to go that way, though!

kirkwaiblinger reviewed (changes requested):
What a great first PR to typescript-eslint! This is very much on the right track, and in a tricky area of the codebase, too. 

Left a few comments and questions.

controversial commented:
Thanks so much for the thoughtful review @kirkwaiblinger !!

I added the test cases from your playground and addressed the missing cases. For the missing case of *unconstrained* type parameters, I added another message to match the one Typescript uses:
> Unsafe type assertion: '{{type}}' could be instantiated with an arbitrary type which could be unrelated to the original type.

---

<details>
<summary>
The only test case that tripped me up a bit was the one you had called <code>parameterExtendsOtherParameter</code>:

</summary>

```ts
function parameterExtendsOtherParameter<T, V extends T>(x: T, y: V) {
  x = y;
  y as T; // allowed

  y = x;
  x as V; // banned; generic arbitrary subtype <-- currently has wrong error message.
}
```

My first impression was that this case should raise the “x is assignable to the constraint of V” message. However, my current implementation instead says “V could be instantiated with an arbitrary type.” **Notably**, though (and somewhat to my surprise) this behavior is consistent with the message that Typescript gives for this case, so I think it’s probably ok as-is?

I [added](https://github.com/typescript-eslint/typescript-eslint/pull/10461/commits/c39c4b72645a86028b46a6837c7c06744eafb54e) a new `parameterExtendsOtherParameter` test where T is *not* unconstrained; here, both TypeScript (for assignment) and my implementation raise “is assignable to the constraint of V” like I expected.

I renamed the original test to <code>parameterExtends<strong>Unconstrained</strong>Parameter</code>; I’m curious to hear your thoughts about whether we should try to produce “assignable to the constraint” for this edge case?

</details>

controversial commented:
Merged the latest `main` to resolve merge conflicts from #10472

When you get a chance, let me know if there’s anything else you need from me @kirkwaiblinger :)

kirkwaiblinger commented:
> My first impression was that this case should raise the “x is assignable to the constraint of V” message. However, my current implementation instead says “V could be instantiated with an arbitrary type.” **Notably**, though (and somewhat to my surprise) this behavior is consistent with the message that Typescript gives for this case, so I think it’s probably ok as-is?

I feel strongly that TS is wrong here, since `V` plainly cannot be _unrelated_ to `T`. It's guaranteed to be a subtype of `T`, since... `V extends T` 🤷 . However, I agree that it makes sense to match their behavior here, and we can follow up by filing an issue to TS if we like. (Feel free to do this if you're interested! Otherwise I probably will at some point 🙂)

> I [added](https://github.com/typescript-eslint/typescript-eslint/pull/10461/commits/c39c4b72645a86028b46a6837c7c06744eafb54e) a new `parameterExtendsOtherParameter` test where T is _not_ unconstrained; here, both TypeScript (for assignment) and my implementation raise “is assignable to the constraint of V” like I expected. I renamed the original test to `parameterExtends**Unconstrained**Parameter`.

🎉

kirkwaiblinger commented on packages/eslint-plugin/tests/rules/no-unsafe-type-assertion.test.ts:
```diff
@@ -1110,6 +1133,247 @@ export function myfunc<CustomObjectT extends string>(
+  x: T,
+  y: V,
+) {
+  y as T; // allowed
```
```suggestion
```

kirkwaiblinger commented on packages/eslint-plugin/tests/rules/no-unsafe-type-assertion.test.ts:
```diff
@@ -1110,6 +1133,247 @@ export function myfunc<CustomObjectT extends string>(
+      {
+        code: `
+function parameterExtendsUnconstrainedParameter<T, V extends T>(x: T, y: V) {
+  y as T; // allowed (as above)
```
```suggestion
```

kirkwaiblinger reviewed (changes requested):
Really outstanding work! Great attention to detail on noticing the additional error message, nicely done on pushing back where it made sense to do so!

Trivial cleanup requested but this is otherwise an enthusiastic approval from me! 🙂 We'll probably leave this open for a little bit as well for another team member to take a look if they like, since it's nuanced stuff. 

Thanks!

<img alt="cuzco" src="https://media3.giphy.com/media/njS7fzNbKflZu/giphy.gif"/>

controversial commented:
cleanup completed! thanks again for your thoughtful reviews, excited to have this merged at some point :)

controversial commented on packages/eslint-plugin/tests/rules/no-unsafe-type-assertion.test.ts:
```diff
@@ -1269,8 +1269,8 @@ function parameterExtendsOtherParameter<T extends string | number, V extends T>(
             endColumn: 9,
-            endLine: 7,
-            line: 7,
+            endLine: 6,
+            line: 6,
```
my bad
